### ISSUE TITLE
Broken IO in multithreaded apps

### ISSUE THREAD

**Issue by plzombie** [open-watcom/open-watcom-v2#791]
Hello. I found what if I call IO functions in separate threads (and more than one thread use IO functions at the same time), this thread crashed. Sometimes it prints error "Semaphore unlocked by wrong owner" or "Thread has no thread-specific data" before crashing.

I created test program what make many threads and each thread write data in separate file. Program was compiled with -bm flag. In very rare cases everything works fine, but usually threads crashed and some of them printed "Thread has no thread-specific data". Here is a [link](https://gist.github.com/plzombie/80d293fc911ba5815928f666f8a65a26) to test program

**Comment by jmalak** [comment]
Visual studio C compiler and Watcom/Open Watcom compilers report some problem with memory.
Do you have enough memory for executable to work with such number of thread and stack size?

**Comment by plzombie** [comment]
I checked again my example and it looks what maximum number of concurrently using descriptors is 64. I updated my gist, now it should work in VC (it even work in MINGW with vc6 runtime). And it works under OW... sometimes. It means OW have same restriction to simultaneously 64 opened files somethere.

I have 16 threaded cpu, if I set number of threads to 20, sometimes (50/50) I got error:
> Thread 0 started
> Thread 2 started
> Thread 1 started
> Thread 7 started
> Thread 2 failed to fill file testdata2.bin
> Semaphore unlocked by wrong owner
> Thread 9 started
> Thread has no thread-specific data
> Thread has no thread-specific data
> Thread has no thread-specific data
> Thread has no thread-specific data
> Thread has no thread-specific data
> Thread has no thread-specific data
> Thread has no thread-specific data
**Semaphore unlocked by wrong owner**
Same error I got in real multithreaded apps.
In rare cases it doesn't even print error and just hang

**Comment by plzombie** [comment]
![image](https://user-images.githubusercontent.com/1690651/147873645-b2342d56-8054-4e66-89bf-ebb02b3c692a.png)
Working fine, print error and hang... 3 in a row

**Comment by jmalak** [comment]
your new sample code doesn't exhibit any problem.
below is directory file listing
```
02.01.2022  13:38           262 144 testdata0.bin
02.01.2022  13:38           262 144 testdata1.bin
02.01.2022  13:38           262 144 testdata10.bin
02.01.2022  13:38           262 144 testdata11.bin
02.01.2022  13:38           262 144 testdata12.bin
02.01.2022  13:38           262 144 testdata13.bin
02.01.2022  13:38           262 144 testdata14.bin
02.01.2022  13:38           262 144 testdata15.bin
02.01.2022  13:38           262 144 testdata16.bin
02.01.2022  13:38           262 144 testdata17.bin
02.01.2022  13:38           262 144 testdata18.bin
02.01.2022  13:38           262 144 testdata19.bin
02.01.2022  13:38           262 144 testdata2.bin
02.01.2022  13:38           262 144 testdata3.bin
02.01.2022  13:38           262 144 testdata4.bin
02.01.2022  13:38           262 144 testdata5.bin
02.01.2022  13:38           262 144 testdata6.bin
02.01.2022  13:38           262 144 testdata7.bin
02.01.2022  13:38           262 144 testdata8.bin
02.01.2022  13:38           262 144 testdata9.bin
```

no problem with any thread.
I attached output for 100 sequential test, no memory problems.
[lst.txt](https://github.com/open-watcom/open-watcom-v2/files/7799072/lst.txt)

**Comment by jmalak** [comment]
I think the source of problem is incorrect use of WaitForMultipleObjects(THREADS_NUM, threads, TRUE, INFINITE); It supposes some synchronization setup for threads.
I added Sleep(1000) after it to have a time to finish all threads and you can see that all treads are finished before program end but for your original code it looks like not all threads are started or finished before program end.
[lst.txt](https://github.com/open-watcom/open-watcom-v2/files/7799095/lst.txt)

**Comment by plzombie** [comment]
No, there is nothing wrong with WaitForMultipleObjects. I created simpliest example. It just reopen files in infinite loop. https://gist.github.com/plzombie/9af6f2aaa8ec7a8b12dbc86e8296fd33
And for me program crashes. I use latest OW build from github

**Comment by jmalak** [comment]
I did a some testing and it looks like some mistake in CRTL, because I get a run-time message "Semaphore unlocked by wrong owner".
I will check locking commands on file operations, it looks like some is unpaired under specific conditions.

**Comment by jmalak** [comment]
Please, could you recheck with latest version of OW V2.
The problem was indirectly caused by bug in Code Generator, issue #805 
Generated code for CRTL was afected by this bug.

**Comment by plzombie** [comment]
Hello. "**Thread has no thread-specific data**" is gone (although I can reproduce it in my first example). But I still get "**Semaphore unlocked by wrong owner**" error.

**Comment by jmalak** [comment]
Thanks for check.
Hm, it is strange. I checked it on my system and I didn't get any problem, except if number of thread is too high then reported a message. I will try to investigate.
Please, what exact host OS version you are using?
I tested it on old 32-bit Windows XP.

**Comment by plzombie** [comment]
Windows 10 x64 21H2, 8core cpu. I will try it on my old system later, maybe I can reproduce

**Comment by jmalak** [comment]
I will try to check it on Windows 10, because OW can use some old construct (from Windows NT 3/4 era) which has problem with Windows 10 WoW64.

**Comment by plzombie** [comment]
I do more tests. I used my [first](https://gist.github.com/plzombie/80d293fc911ba5815928f666f8a65a26) and [last](https://gist.github.com/plzombie/9af6f2aaa8ec7a8b12dbc86e8296fd33) example.
Win10 x64 2c - reproduced
Win10 x86 2c4t - reproduced. For 2nd example dumps occured (see https://disk.yandex.ru/d/Kgi0NdSP6LzYUA, I compiled it with 03.02.2022 build)
ReactOS 0.4.14, x86 1c - first is not reproduced, 2nd is reproduced. Also has stack dumps
I will try to check it also on p3 with winxp later

**Comment by plzombie** [comment]
Tested on WinXP SP3 (p3) - can reproduce 2nd example

**Comment by jmalak** [comment]
I will do more investigation.
Anyway in OW is limit for 64 thread per internal allocated memory block, but it looks like there is some mistake in block handling. it is reason for problems when number of threads are over 64.
